File: history/0.1.2/tatala-client-csharp/src/Example/Tatala/client/ChatRoomClient.cs
using System;
using Example.tatala.proxy;
using QiLeYuan.Tatala.util;
using QiLeYuan.Tools.debug;

namespace Example.tatala.client {
    /// <summary>
    /// This class is a sample for Chat Room c# Client.
    /// Demonstrate long connection server call client directly.
    /// @author JimT
    /// </summary>
    public class ChatRoomClient {
        private static ChatRoomClient chatRoomClient = new ChatRoomClient();
	    private static BlockingQueue messageQueue = new BlockingQueue(10);
	    private ChatRoomClientProxy proxy;
	    private String username;
    	
	    public static ChatRoomClient getInstance(){
		    return chatRoomClient;
	    }

        //want to test on console, you can change MainStart to Main and build as exe.
        public static void MainStart(String[] args) {
            Logging.AddDebug(new ConsoleDebug());
            Logging.setLevel(Logging.Level.INFO);

		    ChatRoomClient room = ChatRoomClient.getInstance();
		    try {
			    room.init();
			    room.login();
                Console.WriteLine("┌-- enter room --┐");
			    while(room.sent()){
				    room.display();
			    }
                Console.WriteLine("└-- exit room --┘");
		    } catch (Exception e) {
                Logging.LogError(e);
		    }
	    }

	    private void init(){
		    proxy = new ChatRoomClientProxy();
	    }
    	
	    private void login(){
            Console.WriteLine("Input user name: ");
		    username = Console.ReadLine();
		    proxy.login(username);
	    }
    	
	    private bool sent(){
		    String myString = Console.ReadLine();
		    proxy.sendMessage(myString);
		    if(myString.Equals("exit")){
			    return false;
		    }
		    myString = username + ": " + myString;
		    messageQueue.Enqueue(myString);
    		
		    return true;
	    }
    	
	    public void receiveMessage(String message){
		    messageQueue.Enqueue(message);
		    display();
	    }
    	
	    private void display(){
		    while(messageQueue.Count > 0){
			    String message = (String)messageQueue.Dequeue();
                Console.WriteLine(message);
		    }
	    }
    	
    }
}

File: tatala-java-example/src/main/java/com/qileyuan/tatala/example/client/ChatRoomClient.java
package com.qileyuan.tatala.example.client;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;

import com.qileyuan.tatala.example.proxy.ChatRoomClientProxy;

/**
 * This class is a sample for Chat Room Client.
 * Demonstrate long connection server call client directly.
 * @author JimT
 *
 */
public class ChatRoomClient {
	private static ChatRoomClient chatRoomClient = new ChatRoomClient();
	private static BlockingQueue<String> messageQueue = new ArrayBlockingQueue<String>(10);
	private ChatRoomClientProxy proxy;
	private String username;
	
	public static ChatRoomClient getInstance(){
		return chatRoomClient;
	}
	
	public static void main(String[] args) {
		ChatRoomClient room = ChatRoomClient.getInstance();
		try {
			room.init();
			room.login();
			System.out.println("^-- enter room --^");

			while(room.sent()){
				room.display();
			}
			
			System.out.println("v-- exit room --v");
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	private void init(){
		proxy = new ChatRoomClientProxy();
	}
	
	private void login() throws IOException{
		System.out.print("Input user name: ");
		BufferedReader buf = new BufferedReader(new InputStreamReader(System.in));
		username = buf.readLine();
		proxy.login(username);
	}
	
	private boolean sent() throws IOException{
		BufferedReader buf = new BufferedReader(new InputStreamReader(System.in));
		String myString = buf.readLine();
		proxy.sendMessage(myString);
		if(myString.equals("exit")){
			return false;
		}
		myString = username + ": " + myString;
		messageQueue.add(myString);
		
		return true;
	}
	
	public void receiveMessage(String message){
		messageQueue.add(message);
		display();
	}
	
	private void display(){
		while(!messageQueue.isEmpty()){
			String message = messageQueue.poll();
			System.out.println(message);
		}
	}
}
